### PR TITLE
[New Rule] Building Block Rule - Attempt to Retrieve User Data from AWS EC2 Instance

### PR DESCRIPTION
**Pull request by terrancedejesus**
## Issues
* https://github.com/elastic/ia-trade-team/issues/358

## Summary
Identifies discovery request `DescribeInstanceAttribute` with the attribute userData and instanceId in AWS CloudTrail
logs. This may indicate an attempt to retrieve user data from an EC2 instance. Adversaries may use this information to
gather sensitive data from the instance or to identify potential vulnerabilities. This is a building block rule that
does not generate an alert on its own, but serves as a signal for anomalous activity.

Screenshot 2024-05-09 at 6 57 03 PM

<details><summary>Example Data</summary>

```
{
  "_index": ".ds-logs-aws.cloudtrail-default",
  "_id": "anonymized-id",
  "_version": 1,
  "_score": 0,
  "_source": {
    "agent": {
      "name": "anonymized-agent-name",
      "id": "anonymized-agent-id",
      "type": "filebeat",
      "ephemeral_id": "anonymized-ephemeral-id",
      "version": "8.13.2"
    },
    "log": {
      "file": {
        "path": "https://anonymized-bucket.s3.region.amazonaws.com/AWSLogs/anonymized-path.json.gz"
      },
      "offset": 9906
    },
    "elastic_agent": {
      "id": "anonymized-agent-id",
      "version": "8.13.2",
      "snapshot": false
    },
    "source": {
      "geo": {
        "continent_name": "North America",
        "region_iso_code": "anonymized-region",
        "city_name": "anonymized-city",
        "country_iso_code": "US",
        "country_name": "United States",
        "region_name": "anonymized-state",
        "location": {
          "lon": -xx.xxxx,
          "lat": xx.xxxx
        }
      },
      "as": {
        "number": 12097,
        "organization": {
          "name": "Anonymized Organization"
        }
      },
      "address": "Anonymized IP",
      "ip": "Anonymized IP"
    },
    "tags": [
      "preserve_original_event",
      "forwarded",
      "aws-cloudtrail"
    ],
    "cloud": {
      "region": "us-east-1"
    },
    "input": {
      "type": "aws-s3"
    },
    "@timestamp": "2024-05-02T19:11:08.000Z",
    "ecs": {
      "version": "8.0.0"
    },
    "related": {
      "user": ["anonymized-user"]
    },
    "data_stream": {
      "namespace": "default",
      "type": "logs",
      "dataset": "aws.cloudtrail"
    },
    "tls": {
      "cipher": "TLS_AES_128_GCM_SHA256",
      "client": {
        "server_name": "anonymized-server-name"
      },
      "version": "1.3",
      "version_protocol": "tls"
    },
    "aws": {
      "s3": {
        "bucket": {
          "name": "anonymized-bucket",
          "arn": "arn:aws:s3:::anonymized-bucket"
        },
        "object": {
          "key": "AWSLogs/anonymized-path.json.gz"
        }
      },
      "cloudtrail": {
        "event_version": "1.09",
        "flattened": {
          "request_parameters": {
            "instanceId": "anonymized-instance-id",
            "attribute": "userData"
          }
        },
        "event_type": "AwsApiCall",
        "read_only": true,
        "user_identity": {
          "access_key_id": "anonymized-access-key",
          "type": "IAMUser",
          "arn": "arn:aws:iam::anonymized-account:user/anonymized-user"
        },
        "event_category": "Management",
        "request_parameters": "anonymized-parameters",
        "request_id": "anonymized-request-id",
        "management_event": true
      }
    },
    "event": {
      "agent_id_status": "verified",
      "ingested": "2024-05-02T19:16:26Z",
      "original": "anonymized-original-event-data",
      "provider": "ec2.amazonaws.com",
      "created": "2024-05-02T19:16:22.606Z",
      "kind": "event",
      "action": "DescribeInstanceAttribute",
      "id": "anonymized-event-id",
      "type": ["info"],
      "dataset": "aws.cloudtrail",
      "outcome": "success"
    },
    "user": {
      "name": "anonymized-user",
      "id": "anonymized-user-id"
    },
    "user_agent": {
      "original": "anonymized-user-agent",
      "name": "aws-sdk-go",
      "device": {
        "name": "Other"
      },
      "version": "1.51.28"
    }
  }
}

```

</details>

### CODE REVIEW

File: rules_building_block/discovery_userdata_request_from_ec2_instance.toml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+rule_id = "c1e79a70-fa6f-11ee-8bc8-f661ea17fbce"
+severity = "low"
+tags = [
+    "Domain: Cloud",
+    "Data Source: AWS",
+    "Data Source: Amazon Web Services",
+    "Data Source: Amazon EC2",
+    "Use Case: Log Auditing",
+    "Tactic: Discovery",
+]
```
Comment (Aegrah) [inline rules_building_block/discovery_userdata_request_from_ec2_instance.toml:38]:
```suggestion
    "Tactic: Discovery",
    "Rule Type: BBR"
]
```

File: rules_building_block/discovery_userdata_request_from_ec2_instance.toml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+updated_date = "2024/04/14"
+
+[rule]
+author = ["Elastic"]
+building_block_type = "default"
+description = """
+Identifies discovery request `DescribeInstanceAttribute` with the attribute userData and instanceId in AWS CloudTrail
+logs. This may indicate an attempt to retrieve user data from an EC2 instance. Adversaries may use this information to
+gather sensitive data from the instance or to identify potential vulnerabilities. This is a building block rule that
+does not generate an alert on its own, but serves as a signal for anomalous activity.
```
Comment (Aegrah) [inline rules_building_block/discovery_userdata_request_from_ec2_instance.toml:16]: If this is the case, is this a higher-order rule? Or are higher-order rules, the rules that query this rule's output?

**Review by Aegrah — APPROVED** [review]

File: rules_building_block/discovery_userdata_request_from_ec2_instance.toml (hunk 1 above)
Comment (Aegrah) [reply rules_building_block/discovery_userdata_request_from_ec2_instance.toml:38]: Should probably add this to unit testing as well.

**Review by Samirbous — APPROVED** [review]

File: rules_building_block/discovery_userdata_request_from_ec2_instance.toml (hunk 2 above)
Comment (terrancedejesus) [reply rules_building_block/discovery_userdata_request_from_ec2_instance.toml:16]: Good question! I'd say this is a good signal for either threat hunting, TDIR or high-order rules but definitely not a high confident indicator of compromise.

**Pull request merged**